Refactor base stages

UdeshyaDhungana opened the pull request:
@rohitpaulk Requesting for an early review.

The refactors are not yet complete in following aspects
- serializer_legacy is still being used in base stages
- Encoder interface is yet to be refactored (hexdump printing part can be incorporated to encoding it seems)

As of now, only major change I've made is in the `decoder/decoder.go` and `api_versions_response.go`.

Also, I found myself duplicating a lot of files, so I'm not sure if I'm correctly following the approach.

rohitpaulk commented on protocol/encoder/encoder.go:
```diff
@@ -0,0 +1,261 @@
+	offset int
+}
+
+func (re *Encoder) Init(raw []byte) {
```
Same note as for `Decoder` - use `NewEncoder`

rohitpaulk commented on protocol/encoder/encoder.go:
```diff
@@ -0,0 +1,261 @@
+	re.offset = 0
+}
+
+func (re *Encoder) PutRawBytesAt(in []byte, offset int, length int) {
```
I don't see any of these used at the moment, let's stick to the functions we need at the moment so it's easier to review along w/ usages

rohitpaulk commented on protocol/encoder/encoder.go:
```diff
@@ -0,0 +1,261 @@
+
+// primitives
+
+func (re *Encoder) PutInt8(in int8) {
```
```suggestion
func (re *Encoder) WriteInt8(in int8) {
```

(Let's use `WriteXYZ` instead of `PutXYZ` - parallel to Decoder's `Read`)

rohitpaulk commented on protocol/encoder/encoder.go:
```diff
@@ -0,0 +1,82 @@
+	offset int
+}
+
+func NewEncoder(bytes []byte) *Encoder {
```
```suggestion
func NewEncoder() *Encoder {
```

Don't think we need `bytes` to be passed in? Could just instantiate an array here

rohitpaulk replied:
```suggestion
func NewEncoder() *Encoder {
```

Don't think we need `bytes` to be passed in? Could just instantiate an array here

rohitpaulk commented on protocol/encoder/encoder.go:
```diff
@@ -0,0 +1,82 @@
+	return re.offset
+}
+
+func (re *Encoder) AllBytes() []byte {
```
The interface for this could be far simpler. I wouldn't expose `Offset` and `AllBytes` - no caller is going to require these values, they'll only need the "written" bytes, and doing `len(bytes)` will give them the offset. 

Would remove these both and make `EncodedBytes` just `Bytes`.

rohitpaulk commented on protocol/encoder/encoder.go:
```diff
@@ -0,0 +1,82 @@
+)
+
+type Encoder struct {
+	bytes  []byte
```
I think you can use bytes.Buffer here and not have to track offset/bytes etc. on your own: https://pkg.go.dev/bytes#Buffer

rohitpaulk commented on protocol/kafkaapi/api_versions_request.go:
```diff
@@ -0,0 +1,46 @@
+
+func (r ApiVersionsRequestBody) encode(enc *encoder.Encoder) {
+
+	if r.Version < 3 {
```
```suggestion
	if r.Version != 3 {
```

(We support exactly one version if I'm not wrong, so let's make an exact check)

rohitpaulk commented on protocol/kafkaapi/api_versions_request.go:
```diff
@@ -0,0 +1,46 @@
+	enc.WriteEmptyTagBuffer()
+}
+
+func (r ApiVersionsRequestBody) Encode() []byte {
```
Let's merge `Encode` and `encode`, don't need the split, doesn't really make it any more readable

rohitpaulk commented on protocol/kafkaapi/api_versions_response.go:
```diff
@@ -0,0 +1,148 @@
+	ThrottleTimeMs int32
+}
+
+func (r *ApiVersionsResponseBody) Decode(d *decoder.Decoder, version int16, logger *logger.Logger, indentation int) (err error) {
```
Doesn't look like we're using `logger` and `indentation` anywhere here.

rohitpaulk commented on protocol/kafkaapi/api_versions_response.go:
```diff
@@ -0,0 +1,148 @@
+		panic("CodeCrafters Internal Error: ApiVersionsResponseBody.Version is not initialized")
+	}
+
+	if err := r.Body.Decode(decoder, r.Body.Version, logger, 1); err != nil {
```
We're validating that `r.Body.Version` is not equal to zero, and then we're passing that same thing into `r.Body.Decode` - doesn't `r.Body.Decode` already have access to `Version`?

rohitpaulk commented on protocol/kafkaapi/api_versions_response.go:
```diff
@@ -0,0 +1,148 @@
+// ApiKeyEntry contains the APIs supported by the broker.
+type ApiKeyEntry struct {
+	// Version defines the protocol version to use for encode and decode
+	Version int16
```
Let's try to get rid of this "Version" field both on `ApiKeyEntry` and `ResponseBody` and move it to the top-level `ApiVersionsResponse`. This isn't actual "state", it's just being passed down for convenience and this can instead be done by passing down a variable to the `Decode` functions.

UdeshyaDhungana replied:
@rohitpaulk , Sure. But it think we should only move it as up as `ResponseBody` because it seems like different body versions can co-exist with same header version.

For eg, both body versions (1 and 2) co-exist with same header version. So, I think the version should be tagged with body instead of the response. I'll remove the versions property and just pass down for any sub-fields of body.

<img width="1088" height="606" alt="Screenshot 2025-09-02 at 13 44 45" src="https://github.com/user-attachments/assets/e54e36c3-ab31-4f59-b14a-917e8319ae3b" />

<img width="1070" height="552" alt="Screenshot 2025-09-02 at 13 44 40" src="https://github.com/user-attachments/assets/1ae9f387-e958-4b3b-81f2-2ecb8987ebd8" />

rohitpaulk commented on protocol/kafkaapi/api_versions_response.go:
Overall this file still feels kind of complex, let's spend some extra time and try to make it as minimal as possible

rohitpaulk commented on protocol/kafkaapi/headers/request_header.go:
```diff
@@ -0,0 +1,34 @@
+}
+
+// encode v2
+func (h RequestHeader) encode(enc *encoder.Encoder) {
```
Let's merge `encode` and `Encode`, the split doesn't really help with readability - also the "encode v2" comment here doesn't make a lot of sense

rohitpaulk commented:
@UdeshyaDhungana let's get the tests passing too (so we can see what the exact changes to fixtures are)

UdeshyaDhungana commented:
- Applied suggested fixes
- Tried to make `api_versions_response.go` a bit cleaner
- Refactored encoder in the same PR
- Log format is now the same as old (with -  and .)
- CI: regenerated fixtures
- Renamed serializer -> `files_manager`
- Base stages only use function for writing metadata and configs and not log dirs.

UdeshyaDhungana commented:
> @UdeshyaDhungana there's too much going on in a single PR here, let's split please - hard to review properly. For example the files_manager stuff can be split into a separate PR

I've removed the `files_manager` part for now. I'll move that into a separate PR.

rohitpaulk commented on protocol/decoder/decoder.go:
```diff
@@ -0,0 +1,271 @@
+		return 0, err
+	}
+
+	if decodedInteger == 0 {
```
Is it intended that here we read a value and don't log it? Also, we're essentially treating the value 1 and 0 the same (here we return 0 directly, below we do `1-1` which is also 0)

rohitpaulk commented on protocol/decoder/decoder.go:
```diff
@@ -0,0 +1,271 @@
+
+	if err != nil {
+		if decodingErr, ok := err.(*errors.PacketDecodingError); ok {
+			return -1, decodingErr.AddContexts("ARRAY_LENGTH", variableName)
```
This `Contexts` thing seems to account for a lot of code here. Would look a lot more simpler if this wasn't needed. I'd think a bit more about exactly what kind of error message we're trying to construct and see if what's needed could be stored as state on `Decoder`. 

Can be a separate PR might need some thinking re: approach

rohitpaulk commented on protocol/kafka_interface/interface.go:
```diff
@@ -0,0 +1,15 @@
+	GetHeader() headers.RequestHeader
+}
+
+type Encodable interface {
```
Let's remove this - isn't used

rohitpaulk commented on protocol/kafkaapi/api_versions_response.go:
```diff
@@ -0,0 +1,147 @@
+	decoder.BeginSubSection("ApiVersions Response")
+
+	defer func() {
+		decoder.EndCurrentSubSection()
```
Let's place this above as `defer decoder.EndCurrentSubSection`, `defer` calls are easier to read when they're placed close to the action they're reversing

rohitpaulk commented on protocol/kafkaapi/api_versions_response.go:
```diff
@@ -0,0 +1,147 @@
+		decoder.EndCurrentSubSection()
+
+		if decodingErr, ok := err.(*errors.PacketDecodingError); ok {
+			detailedError := decodingErr.AddContexts("ApiVersions Response")
```
I think a lot of complexity for this comes from the ordering - `AddContexts` is only done when the error is constructed, instead of before the decoding is done. Might be worth looking into merging the ideas of "SubSection" and "Context" (again, can be separate PR)
